model/modeldecoder/generator/map.go

```diff
@@ -0,0 +1,165 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"fmt"
+	"go/types"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+var mapSupportedTags = []string{tagMaxVals, tagPatternKeys, tagTypesVals, tagRequired}
+
+func generateMapValidation(w io.Writer, fields []structField, f structField, isCustomStruct bool) error {
+	typ, ok := f.Type().Underlying().(*types.Map)
+	if !ok {
+		// this should never happen
+		return errors.New("unexpected field type for tmap")
+	}
```

Comment:
Since this is not production code, I don't mind if this code panics if the assumption is broken. Up to you though.